### PR TITLE
Bluzelle Chain Support

### PR DESCRIPTION
**Pull request by sharish**
## Description

Added support for the Bluzelle blockchain  This PR supports sending funds and staking.

Since bluzelle uses the Cosmos SDK, the cosmos signing logic is reused. The integration is very similar to the Terra and Kava integration.

## Testing instructions

`bootstrap.sh` passes locally. Tests pass, coverage needs met.
`tools/android-test` passes locally. Tests pass, coverage needs met.
`tools/ios-test` passes locally. Tests pass, coverage needs met.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

**Comment by sharish** [comment]
Fixed the breaking android testAddressValidation and added the same in iOS. Verified with a sample run on fork project - [Pull Request #2 - Bluzelle support](https://github.com/frontierdotxyz/wallet-core/pull/2)

**Comment by sharish** [comment]
Asset supported PR requested : https://github.com/trustwallet/assets/pull/9822

**Comment by sharish** [comment]
@catenocrypt @hewigovens : Seems CI machine crashed while building the work-flow. Can you re-run it?

**Comment by hewigovens** [comment]
@sharish will be fixed in this pr: https://github.com/trustwallet/wallet-core/pull/1465

**Comment by hewigovens** [comment]
@sharish could you please rebase or merge master? thanks

**Comment by sharish** [comment]
Yes done. Thanks for the CI fix @hewigovens

File: tests/CoinAddressValidationTests.cpp [hunk 1]
```diff
@@ -330,6 +330,14 @@ TEST(Coin, ValidateAddresBand) {
     EXPECT_FALSE(validateAddress(TWCoinTypeBandChain, "band1pnndgfwsrff86263xzpc5cd3t6yfvgjyqc8000"));
 }
 
+TEST(Coin, ValidateAddresBluzelle) {
```
Comment (catenocrypt) [inline tests/CoinAddressValidationTests.cpp:333]:
```suggestion
TEST(Coin, ValidateAddressBluzelle) {
```

**Review by catenocrypt — CHANGES_REQUESTED** [review]

**Review by ogdeer — APPROVED** [review]

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/bluzelle/TestBluzelleSigner.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import wallet.core.jni.proto.Cosmos
+import wallet.core.jni.proto.Cosmos.SigningOutput
+
+class TestBluzelleSigner {
+
+    init {
+        System.loadLibrary("TrustWalletCore")
+    }
+
+@Test
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/bluzelle/TestBluzelleSigner.kt:29]:
```suggestion
    @Test
```

File: swift/Tests/Blockchains/BluzelleTests.swift [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                  "value": "AlcobsPzfTNVe7uqAAsndErJAjqplnyudaGB0f+R+p3F"
+                },
+                "signature": "0StN9LHr4fIbEzCzyx5bzWJmfgiUqkwoavYGNUDm2shXKmAFogVtOviC4zMNNHKBtFTnHQ07DO1UjqEEWS/2BA=="
+              }
+            ]
+          }
+        }
+        """
+
+        XCTAssertJSONEqual(expectedJSON, output.json)
```
Comment (hewigovens) [inline swift/Tests/Blockchains/BluzelleTests.swift:208]: This is a real world tx?

File: coins.json [hunk 4]
```diff
@@ -953,6 +953,31 @@
...
     }
   },
+  {
+    "id": "bluzelle",
+    "name": "Bluzelle",
+    "coinId": 483,
+    "symbol": "BLZ",
+    "decimals": 6,
+    "blockchain": "Cosmos",
+    "derivationPath": "m/44'/118'/0'/0/0",
```
[the remark this answers is not in the document]
Comment (sharish) [reply coins.json:963]: This was kept in accordance with bluzelle client libraries ([go](https://github.com/bluzelle/blzgo/blob/ab39ab655fa1d91c4b3eb8a594b8776e9d729ab9/client.go#L12) [php](https://github.com/bluzelle/blzphp/blob/467eabc17e42d7f9f0d01d6a2ad691e29dffedb9/src/Utils.php) [python](https://github.com/bluzelle/blzpy/blob/9b3084673913992375a0ae18270cb8edd8165352/lib/bluzelle.py#L19) [ruby](https://github.com/bluzelle/blzrb/blob/957e6dfeeb0b97817908d15249ac55dfad4b980d/lib/bluzelle.rb#L16) [java](https://github.com/bluzelle/blzjava/blob/244f742aef2e9e7d7fb635ab67c8b7534ea4082a/src/com/bluzelle/Bluzelle.java#L86)) where it was extensively used with `m/44'/118'/0'/0/0` derivation path. If this has to be changed, trust derived address will not be in sync with the client libraries. If the derivation path has to be mandated to coinId, we can ask the client libraries to be updated.

File: swift/Tests/Blockchains/BluzelleTests.swift (hunk 3 above)
Comment (sharish) [reply swift/Tests/Blockchains/BluzelleTests.swift:208]: No this is a simulated transaction. For a real world tx

chain-id: net-6
account_number: 590

Has this to be realworld?

Comment (hewigovens) [reply swift/Tests/Blockchains/BluzelleTests.swift:208]: yes, to make sure the code is working

**Comment by hewigovens** [comment]
`TWCoinTypeTests.cpp` is missing, could you please add it?

File: swift/Tests/Blockchains/BluzelleTests.swift (hunk 3 above)
Comment (sharish) [reply swift/Tests/Blockchains/BluzelleTests.swift:131]: This is fixed.

File: coins.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (sharish) [reply coins.json:963]: Sure will change to 483 then.

**Comment by sharish** [comment]
@hewigovens : TWCoinTypeTests.cpp is now added. Thanks

**Review by catenocrypt — APPROVED** [review]

File: coins.json [hunk 5]
```diff
@@ -966,8 +966,8 @@
     "hrp": "bluzelle",
     "explorer": {
       "url": "https://bigdipper.net.bluzelle.com/",
```
Comment (vikmeup) [inline coins.json:968]: can you remove `/` from `url`. but add `/` to `txPath`, `accountPath`

File: coins.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (sharish) [reply coins.json]: @catenocrypt : Changed to 483. Requesting re-review

File: coins.json (hunk 5 above)
Comment (sharish) [reply coins.json]: This is done.

File: coins.json [hunk 6]
```diff
@@ -960,14 +960,14 @@
     "symbol": "BLZ",
     "decimals": 6,
     "blockchain": "Cosmos",
-    "derivationPath": "m/44'/118'/0'/0/0",
+    "derivationPath": "m/44'/483'/0'/0/0",
```
Comment (vikmeup) [inline coins.json:963]: Screen Shot 2021-06-19 at 8 10 08 AM

https://coinmarketcap.com/currencies/bluzelle/

BNT vs BLZ, what's the correct one?

Comment (sharish) [reply coins.json:963]: BNT is a legacy reference and currently used only in internal reference. BLZ will be symbol for enduser prespective.

Comment (sharish) [reply coins.json:963]: We will update [slip's reference](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) soon.

**Review by catenocrypt — APPROVED** [review]

**Comment by iampawan** [comment]
Any update on when it will be merged @vikmeup @catenocrypt ?

**Review by hewigovens — APPROVED** [review]

**Pull request merged**